sky/skylet/providers/gcp/node.py

```diff
@@ -621,6 +631,70 @@ def delete_instance(self, node_id: str, wait_for_operation: bool = True) -> dict
 
         return result
 
+    def resize_disk(
+        self, base_config: dict, instance_name: str, wait_for_operation: bool = True
+    ) -> Tuple[dict, str]:
+        """Resize a Google Cloud disk based on the provided configuration."""
+
+        # Extract the specified disk size from the configuration
+        new_size_gb = base_config["disks"][0]["initializeParams"]["diskSizeGb"]
+
+        # Fetch the instance details to get the disk name and current disk size
+        response = (
+            self.resource.instances()
+            .get(
+                project=self.project_id,
+                zone=self.availability_zone,
+                instance=instance_name,
+            )
+            .execute()
+        )
+        disk_name = response["disks"][0]["source"].split("/")[-1]
+        cur_size_gb = response["disks"][0]["diskSizeGb"]
+
+        # If the new disk size is the same as the existing disk size
+        # resizing is unnecessary
+        if int(new_size_gb) == int(cur_size_gb):
+            # Return empty list instead of raising exception to not break
+            # ray down.
+            return []
+
+        elif int(new_size_gb) < int(cur_size_gb):
+            logger.warning(
+                "Disk resize operation aborted: Specified new disk size (%s GB) is smaller than the current disk size (%s GB)",
+                new_size_gb,
+                cur_size_gb,
+            )
```

Comment:
Could we failed earlier (i.e. before launch)? Maybe check the size of the machine image in `sky/resources.py::_check_image_id`? Also, check if the TPUVM is used?

Reply:
Hi Tian Xia, I also considered failing early in `sky/resources.py::_try_validate_image_id`. This function ensures the image exists and that the `disk_size` provided by the user is larger than the `image_disk_size`. Considering that an image might be associated with multiple instances, each potentially having a different disk size, it seems like we should not trigger a failure here. It appears more appropriate to conduct this check within `gcp/node_provider`, where we have the reusable nodes in `reuse_node_ids`. If my understanding is incorrect, please feel free to point it out.

Reply:
Can you elaborate on "Considering that an image might be associated with multiple instances, each potentially having a different disk size, it seems like we should not trigger a failure here."? I'm not sure what you mean here...
Also, the main reason to do this is that we want to fail before:
* insert a local database record;
* perform a real launch which will incur a charge.
And if we perform the check within the `gcp/node_provider`, it will first launch an instance, try to resize it, and then fail; this will leave a cluster in the `INIT` stage, and the user needs to pay real money.

Reply:
To clarify, GCPTPU does use the same API for disk resizing.